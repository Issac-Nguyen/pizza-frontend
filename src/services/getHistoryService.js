import axios from 'axios';
import config from '../config'

function execute(timeStart, timeEnd) {
    const api ='/api/history?timeStart=' + timeStart + '&timeEnd=' + timeEnd;
    return axios.get(api, {headers: {
        Authorization: "Bearer " + localStorage.getItem(config.keyJwt)
    }})
}

export default execute;