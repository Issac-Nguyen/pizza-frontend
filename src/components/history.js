import React, {useState, useEffect} from 'react';
import {connect} from 'react-redux'
import MustLogin from './mustLogin';
import {startOfDay, endOfDay} from 'date-fns'
import getHistoryService from '../services/getHistoryService';
import {format} from 'date-fns'
import {Row, Col, Collapse, List, message, DatePicker, Divider, Alert, Spin } from 'antd'
import * as _ from 'lodash'
import moment from 'moment';
import {logout} from '../actions/userActions'

const { Panel } = Collapse;

export const History = (props) => {
    let [result, setResult] = useState([])
    let [dateFilter, setDateFilter] = useState(new Date().getTime())
    let [sending, setSending] = useState(false)
    let arrayPanel  = [];
    let resArr = [];

    useEffect(() => {
        if(!props.login)
            return;
        setSending(true)
        getHistoryService(startOfDay(new Date(dateFilter)).getTime(), endOfDay(new Date(dateFilter)).getTime())
        .then(res => {
            
            setSending(false)
            if(!res.data.success) {
                message.error(res.data.msg)
                if(res.data.msg == 'Session timeout. Please re-login') {
                    props.logout();
                }
            } else {
                // console.log(res.data.data)
                resArr = res.data.data;
                let orders = _.groupBy(res.data.data, o => (format(new Date(o.created_at), 'yyyy/MM/dd')));

                _.forEach(orders, (vl, k) => {
                    const vlUniq = _.uniqBy(vl, 'order_id')
                    arrayPanel.push(<Panel header={k} key={k} className="day-header">
                        <Collapse>
                            {vlUniq.map((e, index) => renderNest(index, e))}
                        </Collapse>
                    </Panel>)
                })

                setResult(arrayPanel);
            }
        })
        .catch(err => {
        })
    }, [dateFilter, props.login, props.order])

    

    const renderOrderItem = (orderItems) => {
        return (
            <List
                header={<div>Booked pizzas</div>}
                footer={<div></div>}
                bordered
                dataSource={orderItems}
                renderItem={item => (
                    <Row className="order-order-item">
                        <Col span={6} className="order-order-img"><img src={item.url} /></Col>
                        <Col span={6}>{item.name}</Col>
                        <Col span={6}>${item.price}</Col>
                    </Row>
                )}
    />
        )
    }

    const renderNest = (id, order) => {
        const header = <Row>
            <Col span={5}>
                {order.customer_name}
            </Col>
            <Col span={5}>
                {order.customer_address}
            </Col>
            <Col span={5}>
                {order.customer_phone}
            </Col>
            <Col span={5}>
                Delivery: {order.schedule}
            </Col>
            <Col span={4}>
            {`Total: $${order.total} (Eur ${(order.total * props.rate).toFixed(2)})`}
            </Col>
        </Row>
        const orderItems = resArr.filter(i => i.order_id == order.order_id);
        return (
            <Collapse key={id}>
                <Panel header={header} key="1" className="order-items-panel">
                    {renderOrderItem(orderItems)}
                </Panel>
            </Collapse>
        )
    }

    const disabledDate = (current) => {
        return current < moment().startOf('day');
    }

    const onChangeDate = (date, dateString) => {
        setDateFilter(date.valueOf())
    }

    return (
        <div>
            {!props.login && <MustLogin/>}
            {props.login && <>
            <Row justify="center">
                <Col span="20">
                <Row style={{marginTop: '1rem'}}>
                    <Col span={6}>Choose date history</Col>
                    <Col span={18}>
                    <DatePicker format="YYYY/MM/DD"
                            showToday defaultValue={moment()}
                            onChange={onChangeDate}/>
                    </Col>
                        
                    </Row>
                    <Divider>Your history</Divider>
                    {!!sending && <Spin spinning={sending}/>}
                    {result.length == 0 && 
                    <>
                        <Alert type="info" message={"You don't have any order on " + format(new Date(dateFilter), 'yyyy/MM/dd') }/>
                    </>}
                    {result.length > 0 &&    <Collapse>
                    {result}
                    </Collapse>
                    }
                </Col>
            </Row>
            
            </>}
        </div>
    )
}

const mapStateToProps = state => ({
    login: state.login,
    order: state.order,
    rate: state.rate
})

const mapDispatchToProps = dispatch => ({
    logout: () => {dispatch(logout())}
})

export default connect(mapStateToProps, mapDispatchToProps)(History);