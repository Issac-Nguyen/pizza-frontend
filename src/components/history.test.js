import React from 'react'
import {render, cleanup, act, waitForElementToBeRemoved, getByText, fireEvent} from '@testing-library/react'
import {History} from './history'
import getHistoryService from '../services/getHistoryService'

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  

  // jest.mock('../services/getHistoryService', () => () => Promise.resolve({data: {success: true, data: [{created_at: 1588052042433,
  //   currency: "$",
  //   customer_address: "Ô 18 Lô 903 Hòn Rớ, Phước Đồng",
  //   customer_name: "Nguyen Huu Phat",
  //   customer_phone: "0777955777",
  //   email: "jupitern8@gmail.com",
  //   fee: 2.11,
  //   id: 82,
  //   name: "pizza 3",
  //   order_id: "dd48919a-6982-4f",
  //   pizza_id: 3,
  //   price: 10,
  //   schedule: "immediately",
  //   total: 32.11,
  //   url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"}]}}))

  jest.mock('../services/getHistoryService');

describe('History', () => {
    afterEach(() => {
      jest.clearAllMocks();
      cleanup();
    });

    it('should take a snapshot with login', async () => {
      const result = {data: {success: true, testing: true, data: [{created_at: 1588052042433,
         currency: "$",
         customer_address: "Ô 18 Lô 903 Hòn Rớ, Phước Đồng",
         customer_name: "Nguyen Huu Phat",
         customer_phone: "0777955777",
         email: "jupitern8@gmail.com",
         fee: 2.11,
         id: 82,
         name: "pizza 3",
         order_id: "dd48919a-6982-4f",
         pizza_id: 3,
         price: 10,
         schedule: "immediately",
         total: 32.11,
         url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"}]}};
       const mockFn = getHistoryService.mockResolvedValueOnce(result);
                const {asFragment, getByText, container} = (render(<History login={true}/>));
           await waitForElementToBeRemoved(() => container.querySelector('.ant-spin'));
           fireEvent.click(container.querySelector('.day-header'))
           expect(asFragment(<History login={true}/>)).toMatchSnapshot()
       
    })
    
    it('should take a snapshot without login', () => {
        const { asFragment } = render(<History login={false}/>)
        expect(asFragment(<History login={false}/>)).toMatchSnapshot()
     })
 
     
})