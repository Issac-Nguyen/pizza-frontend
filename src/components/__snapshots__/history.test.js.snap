// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History should take a snapshot with login 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-row ant-row-center"
    >
      <div
        class="ant-col ant-col-20"
      >
        <div
          class="ant-row"
          style="margin-top: 1rem;"
        >
          <div
            class="ant-col ant-col-6"
          >
            Choose date history
          </div>
          <div
            class="ant-col ant-col-18"
          >
            <div
              class="ant-picker"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  placeholder="Select date"
                  readonly=""
                  size="12"
                  title="2020/04/29"
                  value="2020/04/29"
                />
                <span
                  class="ant-picker-suffix"
                >
                  <span
                    aria-label="calendar"
                    class="anticon anticon-calendar"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="calendar"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                      />
                    </svg>
                  </span>
                </span>
                <span
                  class="ant-picker-clear"
                >
                  <span
                    aria-label="close-circle"
                    class="anticon anticon-close-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="close-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-divider ant-divider-horizontal ant-divider-with-text-center"
          role="separator"
        >
          <span
            class="ant-divider-inner-text"
          >
            Your history
          </span>
        </div>
        <div
          class="ant-collapse ant-collapse-icon-position-left"
        >
          <div
            class="ant-collapse-item day-header"
          >
            <div
              aria-expanded="false"
              class="ant-collapse-header"
              role="button"
              tabindex="0"
            >
              <span
                aria-label="right"
                class="anticon anticon-right ant-collapse-arrow"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
              2020/04/28
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`History should take a snapshot without login 1`] = `
<DocumentFragment>
  <div>
    <h1>
      You need to login to see this feature
    </h1>
  </div>
</DocumentFragment>
`;
